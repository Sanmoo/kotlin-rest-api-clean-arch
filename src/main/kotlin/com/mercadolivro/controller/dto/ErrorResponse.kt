package com.mercadolivro.controller.dto

data class ErrorResponse(
    val httpCode: Int,
    val message: String,
    val internalCode: String,
    val errors: List<FieldErrorResponse>?,
)
