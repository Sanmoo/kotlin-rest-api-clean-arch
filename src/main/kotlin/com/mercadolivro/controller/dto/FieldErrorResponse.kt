package com.mercadolivro.controller.dto

data class FieldErrorResponse (
    val message: String,
    val field: String
)
