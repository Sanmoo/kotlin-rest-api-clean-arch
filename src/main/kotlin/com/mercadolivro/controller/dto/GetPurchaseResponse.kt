package com.mercadolivro.controller.dto

import java.math.BigDecimal
import java.time.LocalDateTime

data class GetPurchaseResponse(
    val id: Int,
    val customerId: Int,
    val bookIds: Set<Int>,
    val nfe: String?,
    val price: BigDecimal,
    val createdAt: LocalDateTime
)
