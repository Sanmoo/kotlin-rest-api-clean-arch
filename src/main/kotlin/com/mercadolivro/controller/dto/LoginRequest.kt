package com.mercadolivro.controller.dto

data class LoginRequest(
    val email: String,
    val password: String
)
