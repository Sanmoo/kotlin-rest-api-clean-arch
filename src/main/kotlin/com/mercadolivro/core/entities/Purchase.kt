package com.mercadolivro.core.entities

import java.math.BigDecimal
import java.time.LocalDateTime

data class Purchase(
    override val id: Int,
    val customerId: Int,
    val bookIds: Set<Int>,
    val nfe: String?,
    val price: BigDecimal,
    val createdAt: LocalDateTime
) : Entity
