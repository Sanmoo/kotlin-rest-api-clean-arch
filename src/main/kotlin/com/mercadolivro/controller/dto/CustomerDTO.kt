package com.mercadolivro.controller.dto

import com.mercadolivro.core.entities.CustomerStatus
import javax.validation.constraints.NotBlank

data class CustomerDTO(
    val id: Int?,

    @field:NotBlank(message = "Must not be empty")
    val name: String,

    @field:NotBlank
    val email: String,

    val status: CustomerStatus?
)
