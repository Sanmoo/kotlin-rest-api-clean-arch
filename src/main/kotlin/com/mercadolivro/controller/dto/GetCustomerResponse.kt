package com.mercadolivro.controller.dto

import com.mercadolivro.core.entities.CustomerStatus

data class GetCustomerResponse(
    val id: Int,
    val name: String,
    val email: String,
    val status: CustomerStatus
)
