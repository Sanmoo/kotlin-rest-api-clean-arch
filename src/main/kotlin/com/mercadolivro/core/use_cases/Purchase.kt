package com.mercadolivro.core.use_cases

import com.mercadolivro.core.entities.Book
import com.mercadolivro.core.entities.BookStatus
import com.mercadolivro.core.entities.Customer
import com.mercadolivro.core.entities.Purchase
import com.mercadolivro.core.use_cases.exceptions.Errors
import com.mercadolivro.core.use_cases.ports.AsynchronousCoordinator
import com.mercadolivro.core.use_cases.ports.BookRepository
import com.mercadolivro.core.use_cases.ports.CustomerRepository
import com.mercadolivro.core.use_cases.ports.PurchaseRepository
import java.math.BigDecimal
import java.time.LocalDateTime
import java.util.*

class Purchase(
    private val purchaseRepository: PurchaseRepository,
    private val bookRepository: BookRepository,
    private val customerRepository: CustomerRepository,
    private val async: AsynchronousCoordinator,
) {
    data class Input(val customerId: Int, val bookIds: Set<Int>)

    data class Output(
        val id: Int,
        val customer: GetCustomerDetails.Output,
        val books: List<GetBookDetails.Output>,
        val nfe: String? = null,
        val price: BigDecimal,
        val createdAt: LocalDateTime
    ) {
        companion object {
            fun fromEntities(p: Purchase, c: Customer, bs: List<Book>): Output {
                return Output(
                    id = p.id,
                    customer = GetCustomerDetails.Output.fromCustomerEntity(c),
                    books = bs.map { GetBookDetails.Output.fromBookEntity(it) },
                    price = p.price,
                    createdAt = p.createdAt
                )
            }
        }
    }

    fun purchase(input: Input): Output {
        val books = bookRepository.getAllByIds(input.bookIds)

        if (books.size != input.bookIds.size) {
            val foundBookIds = books.map { it.id }
            throw Errors.ML103.toResourceNotFoundException(input.bookIds.subtract(foundBookIds))
        }

        if (books.any { it.status in listOf(BookStatus.SOLD, BookStatus.DELETED) }) {
            throw Errors.ML104.toOperationNotAllowed()
        }

        val customer = customerRepository.getById(input.customerId) ?: throw Errors.ML201.toResourceNotFoundException(input.customerId)

        val purchaseRepositoryCreateInput = PurchaseRepository.CreateInput(
            customerId = customer.id,
            price = books.sumOf { it.price },
            bookIds = input.bookIds
        )

        val purchase = purchaseRepository.create(purchaseRepositoryCreateInput)

        async.doAsync {
            val updatedPurchase = purchase.copy(nfe = UUID.randomUUID().toString())
            purchaseRepository.update(updatedPurchase.id, updatedPurchase)
            books.forEach {
                bookRepository.update(it.id, it.copy(status = BookStatus.SOLD))
            }
        }

        return Output.fromEntities(purchase, customer, books)
    }
}
