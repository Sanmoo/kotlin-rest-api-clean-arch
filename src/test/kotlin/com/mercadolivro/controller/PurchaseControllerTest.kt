package com.mercadolivro.controller

import au.com.origin.snapshots.Expect
import au.com.origin.snapshots.junit5.SnapshotExtension
import com.fasterxml.jackson.databind.ObjectMapper
import com.mercadolivro.controller.dto.PostPurchaseRequest
import com.mercadolivro.core.entities.TestFactories.Companion.makeBook
import com.mercadolivro.core.entities.TestFactories.Companion.makeCustomer
import com.mercadolivro.core.entities.TestFactories.Companion.makePurchase
import com.mercadolivro.core.use_cases.Purchase
import com.ninjasquad.springmockk.MockkBean
import io.mockk.every
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.http.MediaType
import org.springframework.security.test.context.support.WithMockUser
import org.springframework.test.context.ActiveProfiles
import org.springframework.test.context.ContextConfiguration
import org.springframework.test.web.servlet.MockMvc
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders
import org.springframework.test.web.servlet.result.MockMvcResultMatchers

@SpringBootTest
@AutoConfigureMockMvc
@ContextConfiguration
@ActiveProfiles("test")
@WithMockUser
@ExtendWith(SnapshotExtension::class)
internal class PurchaseControllerTest {
    @MockkBean
    private lateinit var purchase: Purchase

    @Autowired
    private lateinit var mockMvc: MockMvc

    @Autowired
    private lateinit var objectMapper: ObjectMapper

    private lateinit var expect: Expect

    @Test
    fun `it purchases a book`() {
        // Arrange
        val customer = makeCustomer(1)
        val book1 = makeBook(1)
        val book2 = makeBook(2)
        val bookIds = setOf(book1.id, book2.id)
        val purchaseEntity = makePurchase(1).copy(customerId = customer.id, bookIds = bookIds)
        every {
            purchase.purchase(
                Purchase.Input(customerId = customer.id, bookIds = bookIds)
            )
        } returns Purchase.Output.fromEntities(purchaseEntity, customer, listOf(book1, book2))
        val purchaseRequest = PostPurchaseRequest(
            customerId = customer.id,
            bookIds = bookIds
        )
        val request = MockMvcRequestBuilders.post("/purchases").contentType(MediaType.APPLICATION_JSON)
            .content(objectMapper.writeValueAsString(purchaseRequest))

        // Act
        val result = mockMvc.perform(request)
            .andExpect(MockMvcResultMatchers.status().isCreated)
            .andReturn()

        // Assert
        expect.toMatchSnapshot(result.response.contentAsString)
    }
}