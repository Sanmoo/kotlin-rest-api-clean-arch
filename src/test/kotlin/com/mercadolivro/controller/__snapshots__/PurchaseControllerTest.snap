com.mercadolivro.controller.PurchaseControllerTest.it purchases a book=[
{"id":1,"customerId":1,"bookIds":[1,2],"price":0.05,"createdAt":"2022-03-05T18:26:16"}
]